src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -546,4 +617,197 @@ private CompletableFuture<List<V>> setToValueCache(List<V> assembledValues, List
     private static <T> DispatchResult<T> emptyDispatchResult() {
         return (DispatchResult<T>) EMPTY_DISPATCH_RESULT;
     }
+
+    private class DataLoaderSubscriber implements Subscriber<V> {
+
+        private final CompletableFuture<List<V>> valuesFuture;
+        private final List<K> keys;
+        private final List<Object> callContexts;
+        private final List<CompletableFuture<V>> queuedFutures;
+
+        private final List<K> clearCacheKeys = new ArrayList<>();
+        private final List<V> completedValues = new ArrayList<>();
+        private int idx = 0;
+        private boolean onErrorCalled = false;
+        private boolean onCompleteCalled = false;
+
+        private DataLoaderSubscriber(
+            CompletableFuture<List<V>> valuesFuture,
+            List<K> keys,
+            List<Object> callContexts,
+            List<CompletableFuture<V>> queuedFutures
+        ) {
+            this.valuesFuture = valuesFuture;
+            this.keys = keys;
+            this.callContexts = callContexts;
+            this.queuedFutures = queuedFutures;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            subscription.request(keys.size());
+        }
+
+        // onNext may be called by multiple threads - for the time being, we pass 'synchronized' to guarantee
+        // correctness (at the cost of speed).
+        @Override
+        public synchronized void onNext(V value) {
+            assertState(!onErrorCalled, () -> "onError has already been called; onNext may not be invoked.");
+            assertState(!onCompleteCalled, () -> "onComplete has already been called; onNext may not be invoked.");
+
+            K key = keys.get(idx);
+            Object callContext = callContexts.get(idx);
+            CompletableFuture<V> future = queuedFutures.get(idx);
+            if (value instanceof Try) {
+                // we allow the batch loader to return a Try so we can better represent a computation
+                // that might have worked or not.
+                Try<V> tryValue = (Try<V>) value;
+                if (tryValue.isSuccess()) {
+                    future.complete(tryValue.get());
+                } else {
+                    stats.incrementLoadErrorCount(new IncrementLoadErrorCountStatisticsContext<>(key, callContext));
+                    future.completeExceptionally(tryValue.getThrowable());
+                    clearCacheKeys.add(keys.get(idx));
+                }
+            } else {
+                future.complete(value);
+            }
+
+            completedValues.add(value);
+            idx++;
```

Comment:
I think we might want another call to `subscription.request()` here just in case??  I mean I know we asked for `key.size()` earlier.  You are allowed to ask for more than there are... Hmmm....